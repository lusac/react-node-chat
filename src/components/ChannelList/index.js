import React from 'react'
import PropTypes from 'prop-types'
import NewChannelModal from '../NewChannelModal'

export default class ChannelList extends React.Component {
  constructor() {
    super()
    this.state = {
      showModal: false
    }
  }

  selectChat(id) {
    this.props.socket.emit('join channel', id)
  }

  toggleModal() {
    this.setState({
      showModal: !this.state.showModal
    })
  }

  render() {
    return (
      <div className="channel-list">
        <div className="channel-list__header">
          Ol√° <strong className="channel-list__header__name">{this.props.username}</strong>!
        </div>

        <div className="channel-list__section">
          <div className="channel-list__section__title">
            <span>Canais</span>
            <span
              className="plus"
              onClick={this.toggleModal.bind(this)}>+</span>
          </div>

          {Object.keys(this.props.channels).map(id => {
            return (
              <span
                key={id}
                className={"channel-list__item " + (this.props.channel.id === id ? 'selected' : '')}
                onClick={this.selectChat.bind(this, id)}>
                  # {this.props.channels[id].name}</span>
            )
          })}
        </div>

        {this.state.showModal &&
          <NewChannelModal
            toggleModal={this.toggleModal.bind(this)}
            socket={this.props.socket} />
        }
      </div>
    )
  }
}

ChannelList.propTypes = {
  socket: PropTypes.object,
  channel: PropTypes.object,
  username: PropTypes.string,
  channels: PropTypes.object,
};