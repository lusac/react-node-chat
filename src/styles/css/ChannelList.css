.channel-list {
  display: grid;
  height: 100vh;
  background: #f2f2f2;
  grid-template-rows: 70px auto;
}
.channel-list__header {
  display: flex;
  align-items: center;
  justify-content: center;
  border-right: 1px solid #e1e1e1;
}
.channel-list__header__name {
  color: #00c876;
  max-width: 110px;
  overflow: hidden;
  font-weight: bold;
  margin: 0 0 0 5px;
  white-space: nowrap;
  text-overflow: ellipsis;
}
.channel-list__section {
  display: flex;
  overflow: auto;
  flex-direction: column;
}
.channel-list__section__title {
  padding: 18px;
  border-right: 1px solid #e1e1e1;
}
.channel-list__item {
  cursor: pointer;
  padding: 6px 18px;
  border-right: 1px solid #e1e1e1;
}
.channel-list__item:hover {
  background: #EFEFEF;
}
.channel-list__item.selected {
  background: #FFF;
  border-right: none;
}

/*# sourceMappingURL=ChannelList.css.map */
