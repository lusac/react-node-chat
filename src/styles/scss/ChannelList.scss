.channel-list {
  display: grid;
  height: 100vh;
  background: #f2f2f2;
  grid-template-rows: 30px auto;

  &__header {
    display: flex;
    align-items: center;
    justify-content: center;
    border-right: 1px solid #e1e1e1;
  }

  &__section {
    overflow: auto;
    display: flex;
    flex-direction: column;

    &__title {
      padding: 18px;
      border-right: 1px solid #e1e1e1;
    }
  }

  &__item {
    cursor: pointer;
    padding: 6px 18px;
    border-right: 1px solid #e1e1e1;

    &:hover {
      background: #EFEFEF;
    }

    &.selected {
      background: #FFF;
      border-right: none;
    }
  }
}