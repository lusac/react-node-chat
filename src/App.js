/* global process navigator document Notification */
import React, { Component } from 'react'
import './styles/css/index.css'
import Channel from './components/Channel'
import ChannelList from './components/ChannelList'
import { Header } from './components/Header'
import socketIOClient from 'socket.io-client'

class App extends Component {
  constructor() {
    super()
    this.state = {
      user: {
        channels: []
      },
      channels: {},
      channel: {},
      notifications: {}
    }
  }

  connectToSocket() {
    const host = process.env.NODE_ENV === 'production' ? '' : 'http://localhost:3001'
    let socket = socketIOClient(host)

    // socket.on('connect', () => {});

    socket.on('channels', (channels) => {
      this.setState({
        channels: channels
      })
    })

    socket.on('got channel', (channel) => {
      this.setState({
        channel: channel,
        notifications: {
          ...this.state.notifications,
          [channel.id]: 0
        }
      })
    })

    socket.on('leaved channel', (channel) => {
      this.setState({
        channel: channel,
        user: {
          ...this.state.user,
          channels: this.state.user.channels.filter((id) => {
            return id !== channel.id
          })
        },
        notifications: {
          ...this.state.notifications,
          [channel.id]: 0
        }
      })
    })

    socket.on('joined channel', (channel) => {
      this.setState({
        channel: channel,
        user: {
          ...this.state.user,
          channels: [
            ...this.state.user.channels,
            channel.id
          ]
        }
      })
    })

    socket.on('message', (data) => {
      this.pushNotification(data)

      if (data.channelID === this.state.channel.id) {
        this.handleMessageFromCurrentChannel(data)
      } else {
        this.handleMessageFromOtherChannel(data)
      }
    })

    this.setState({
      socket: socket
    })
  }

  pushNotification(data) {
    if (!document.hasFocus() && Notification.permission !== 'denied') {
      navigator.serviceWorker.ready.then((registration) => {
        registration.showNotification(this.state.channels[data.channelID].name, {
          body: `${data.msg.username}: ${data.msg.text}`,
          vibrate: [200, 100, 200, 100, 200, 100, 200],
          tag: 'vibration-sample'
        })
      })
    }
  }

  handleMessageFromCurrentChannel(data) {
    this.setState({
      channel: {
        ...this.state.channel,
        msgs: [
          ...this.state.channel.msgs,
          data.msg
        ]
      }
    })
  }

  handleMessageFromOtherChannel(data) {
    this.setState({
      notifications: {
        ...this.state.notifications,
        [data.channelID]: this.state.notifications[data.channelID] + 1 || 1
      }
    })
  }

  onUsernameSubmit(e) {
    e.preventDefault()
    this.setState({
      user: {
        ...this.state.user,
        name: this.draftUsername,
        color: Math.floor((Math.random() * 6) + 1)
      }
    }, () => {
      this.connectToSocket()
    })
  }

  onDraftUsernameChange(e) {
    this.draftUsername = e.target.value
  }

  renderChooseName() {
    if (!this.state.user.name) {
      return (
        <div className="form-choose-name">
          <div className="form-choose-name__content">
            <form onSubmit={this.onUsernameSubmit.bind(this)}>
              <label>Escolha um apelido</label>
              <input
                type="text"
                value={this.draftUsername}
                placeholder="ex: lucas.cruz"
                onChange={this.onDraftUsernameChange.bind(this)} />
              <button className="blue">entrar</button>
            </form>
          </div>
        </div>
      )
    }
  }

  renderChat() {
    if (this.state.user.name) {
      return (
        <div>
          <Header />
          <input type="checkbox" id="toggleChannelList"></input>
          <div className="chat">
            <ChannelList
              user={this.state.user}
              channel={this.state.channel}
              channels={this.state.channels}
              socket={this.state.socket}
              notifications={this.state.notifications} />
            <Channel
              user={this.state.user}
              channel={this.state.channel}
              socket={this.state.socket}  />
          </div>
        </div>
      )
    }
  }

  render() {
    return (
      <div className="App">
        {this.renderChooseName()}
        {this.renderChat()}
      </div>
    )
  }
}

export default App