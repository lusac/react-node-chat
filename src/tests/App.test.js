/* global it document describe expect beforeEach */
import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme'
import renderer from 'react-test-renderer'
import Adapter from 'enzyme-adapter-react-16'
import App from '../App';
import ChannelList from '../components/ChannelList';

configure({ adapter: new Adapter() })

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('Snapshot', () => {
  it('Initial', () => {
    const renderedValue =  renderer.create(<App />).toJSON()
    expect(renderedValue).toMatchSnapshot()
  })
})

describe('Rendered', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(<App />)
  })

  it('should render chat after setting the username', () => {
    wrapper.instance().draftUsername = 'oie'
    wrapper.find('.form-choose-name form').at(0).simulate('submit', { preventDefault: () => {}, target: {} })
    expect(wrapper.find('ChannelList').at(0).text()).toBe('<ChannelList />')
    expect(wrapper.find('Channel').at(0).text()).toBe('<Channel />')
  })
})
