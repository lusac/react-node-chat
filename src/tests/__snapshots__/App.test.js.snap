// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Initial 1`] = `
<div
  className="App"
>
  <div
    className="form-choose-name"
  >
    <div
      className="form-choose-name__content"
    >
      <form
        onSubmit={[Function]}
      >
        <label>
          Escolha um apelido
        </label>
        <input
          onChange={[Function]}
          placeholder="ex: lucas.cruz"
          type="text"
          value={undefined}
        />
        <button
          className="blue"
        >
          entrar
        </button>
      </form>
    </div>
  </div>
</div>
`;
